clang/include/clang/AST/DeclTemplate.h

```diff
@@ -1533,13 +1535,11 @@ class NonTypeTemplateParmDecl final
   /// Return the constraint introduced by the placeholder type of this non-type
   /// template parameter (if any).
   Expr *getPlaceholderTypeConstraint() const {
-    return hasPlaceholderTypeConstraint() ? *getTrailingObjects<Expr *>() :
-        nullptr;
+    return PlaceholderTypeConstraintInitialized ? *getTrailingObjects<Expr *>()
+                                                : nullptr;
   }
 
-  void setPlaceholderTypeConstraint(Expr *E) {
-    *getTrailingObjects<Expr *>() = E;
```

Comment:
While I like the assert-safety that we get from the `PlaceholderTypeConstraintInitialized`, I'm not sure it is worth the rigamarole.  Why not just have the `ctor` do `setPlaceholderTypeConstraint`?

Reply:
Both inside `Sema` and `ASTReaderDecl` the `NonTypeTemplateParmDecl` is created relatively far from the parsing / deserialization of the constraint.

Reply:
I'm not complaining about the 'set' method, just that the concern is that it is uninitialized.  I'm saying, why not just initialize it to nullptr?  Or am I misunderstanding the concern.

Reply:
Ok, then we're on the same page.  I do not see a difference between "explicitly null" and "not initialized" here.  I think that is a differentiation without a cause.